tools/scripts/list_stl_dependencies.sh

```diff
@@ -22,30 +22,74 @@
 #   ./list_stl_dependencies.sh
 #=====================================================
 
+SCOPE=${1:-list}
 COMPONENTS=(iceoryx_hoofs iceoryx_posh)
 SOURCE_DIR=(source include)
 WORKSPACE=$(git rev-parse --show-toplevel)
+QNX_PLATFORM_DIR=$WORKSPACE/iceoryx_hoofs/platform/qnx/
+USELIST=$WORKSPACE/tools/scripts/used-headers.txt
+TEMP_FILE=$(mktemp)
+GET_HEADER_NAME="\<\K[^<>]+(?=>)" # Matches the content between angle brackets
 
 for COMPONENT in ${COMPONENTS[@]}; do
     for DIR in ${SOURCE_DIR[@]}; do
         GREP_PATH="${GREP_PATH} ${WORKSPACE}/${COMPONENT}/$DIR"
     done
 done
 
+echo "# QNX platform/ libc headers" >> $TEMP_FILE
+
+echo "# usage of headers for QNX"
+gcc -w -fpreprocessed -dD -E $(find $QNX_PLATFORM_DIR -type f -iname *.cpp -o -iname *.hpp) \
+ | grep -e "#include <" \
+ | grep -oP $GET_HEADER_NAME \
+ | sort \
+ | uniq \
+ | tee -a $TEMP_FILE \
+ | cat
+
+
+echo "# iceoryx_posh / iceoryx_hoofs headers" >> $TEMP_FILE
+# GCC can't preprocess .inl so we grep them in plain text
 echo
-echo usage of std components by file
-grep -RIne "std::" $GREP_PATH | sed -n  "s/\([^:]*\:[0-9]*\)\:.*\(std::[a-zA-Z_]*\).*/\ \ \1  \2/p" | sort | uniq
+echo "# usage of <..> header includes"
+{ gcc -w -fpreprocessed -dD -E $(find $GREP_PATH -type f -iname *.cpp -o -iname *.hpp); cat $(find $GREP_PATH -type f -iname *.inl); } \
+ | grep -e "#include <" \
+ | grep -oP $GET_HEADER_NAME \
+ | sort \
+ | uniq \
+ | tee -a $TEMP_FILE \
+ | cat
+
+if [[ "$SCOPE" == "check" ]]; then
+    echo
+    echo "Comparing the used system headers against the list.."
+    diff $TEMP_FILE $USELIST
+    if [ $? -eq 1 ]; then
+        echo "One or more header is not on the list, please remove the header usage or extend the list!"
+        exit 1
+    fi
+    echo "No header divergence found!"
+fi
 
 echo
-echo files with stl dependency
+echo "# usage of std components"
+{ gcc -w -fpreprocessed -dD -E $(find $GREP_PATH -type f -iname *.cpp -o -iname *.hpp); cat $(find $GREP_PATH -type f -iname *.inl); } \
+ | grep -HIne "std::" \
+ | sed -n  "s/\([^:]*\:[0-9]*\)\:.*\(std::[a-zA-Z_]*\).*/\ \  \2/p" \
+ | sort \
+ | uniq
+
+echo
+echo "# files with stl dependency"
 grep -RIne "std::" $GREP_PATH | sed -n  "s/\([^:]*\)\:[0-9]*\:.*std::[a-zA-Z_]*.*/\1/p" | xargs -I{} basename {} | sort | uniq
 
 echo
-echo using namespace with std component
+echo "# using namespace with std component"
 grep -RIne ".*using[ ]*namespace[ ]*std" $GREP_PATH | sed -n "s/\(.*\)/\ \ \1/p"
 
 echo
-echo usage of std components
+echo "# usage of std components"
 grep -RIne "std::" $GREP_PATH | sed -n  "s/.*\(std::[a-zA-Z_]*\).*/\ \ \1/p" | sort | uniq
```

Comment:
I have no idea what `sed` does and what to expect as output. For maintainability, please add a comment what one should expect as output.

Btw, this also prints `std::` without any component. Does this mean that the `sed` command is removing too much?

Reply:
AFAIK this just extracts all `std::` with all comments. As it was written by @elfenpiff I'm passing the question along. As of today this is just for informational purposes, if we want to run it on the CI we need to preprocess the file and removing the comments.